evm_loader/program/src/solana_backend.rs

```diff
@@ -1,118 +1,44 @@
 use evm::{
-    backend::{Basic, Backend, Apply, Log},
+    backend::{Basic, Backend},
     CreateScheme, Capture, Transfer, ExitReason
 };
 use core::convert::Infallible;
 use primitive_types::{H160, H256, U256};
 use sha3::{Digest, Keccak256};
 use solana_sdk::{
-    account_info::{next_account_info, AccountInfo},
+    account_info::AccountInfo,
     pubkey::Pubkey,
-    program_error::ProgramError,
-    sysvar::{clock::Clock, Sysvar},
+    program::invoke_signed,
     instruction::{Instruction, AccountMeta},
 };
-use std::{
-    cell::RefCell,
-};
-
-use crate::solidity_account::SolidityAccount;
-use crate::account_data::AccountData;
-use solana_sdk::program::invoke;
-use solana_sdk::program::invoke_signed;
+use std::cell::Ref;
 use std::convert::TryInto;
-use arrayref::{array_ref, array_refs, array_mut_ref, mut_array_refs};
-
-fn keccak256_digest(data: &[u8]) -> H256 {
-    H256::from_slice(Keccak256::digest(&data).as_slice())
-}
-
-pub fn solidity_address<'a>(key: &Pubkey) -> H160 {
-    H256::from_slice(key.as_ref()).into()
+use arrayref::{array_ref, array_refs};
+
+pub trait AccountStorage {
+    fn origin(&self) -> H160;
+    fn get_account_solana_address(&self, address: H160) -> Option<&Pubkey>;
+    fn get_contract_seeds(&self) -> Option<(H160, u8)>;
+    fn get_caller_seeds(&self) -> Option<(H160, u8)>;
+    fn exists(&self, address: H160) -> bool;
+    fn basic(&self, address: H160) -> Basic;
+    fn code_hash(&self, address: H160) -> H256;
+    fn code_size(&self, address: H160) -> usize;
+    fn code(&self, address: H160) -> Vec<u8>;
+    fn storage(&self, address: H160, index: H256) -> H256;
```

Comment:
These methods should be in SolidityAccount. If you move them there, there will be much less duplicate code.
The `AccountStorage` should only have a method for getting the `SolidityAccount` by the Ethereum address.